include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,461 @@
+        std::swap(lhs._coro, rhs._coro);
+    }
+
+    async_generator& operator=(async_generator&& other) noexcept {
+        if (this == &other) {
+            return *this;
+        }
+        if (_coro) {
+            _coro.destroy();
+        }
+        _coro = std::exchange(other._coro, nullptr);
+        return *this;
+    }
+
+    [[nodiscard]] auto begin() noexcept {
+        using base_awaiter = internal::next_awaiter<yielded_type>;
+        class begin_awaiter final : public base_awaiter {
+            using base_awaiter::_promise;
+
+        public:
+            explicit begin_awaiter(std::nullptr_t) noexcept
+                : base_awaiter{nullptr}
+            {}
+            explicit begin_awaiter(handle_type producer_coro) noexcept
+                : base_awaiter{producer_coro.promise(), producer_coro}
+            {}
+            bool await_ready() const noexcept {
+                return _promise == nullptr || base_awaiter::await_ready();
+            }
+
+            iterator await_resume() {
+                if (_promise == nullptr) {
+                    return iterator{nullptr};
+                }
+                if (_promise->finished()) {
+                    _promise->rethrow_if_unhandled_exception();
+                    return iterator{nullptr};
+                }
+                return iterator{
+                    handle_type::from_promise(*static_cast<promise_type *>(_promise))
+                };
+            }
+        };
+
+        if (_coro) {
+            return begin_awaiter{_coro};
+        } else {
+            return begin_awaiter{nullptr};
+        }
+    }
+
+    [[nodiscard]] std::default_sentinel_t end() const noexcept {
+        return {};
+    }
+};
+
+template <typename Ref, typename Value>
+class async_generator<Ref, Value>::promise_type final : public internal::async_generator_promise_base<yielded_type> {
+public:
+    async_generator get_return_object() noexcept {
+        return async_generator{*this};
+    }
+
+    yielded_type value() const noexcept {
+        return static_cast<yielded_type>(*this->_value);
+    }
+};
+
+template <typename Ref, typename Value>
+class async_generator<Ref, Value>::iterator final {
+private:
+    using handle_type = async_generator::handle_type;
+    handle_type _coro = nullptr;
+
+public:
+    using iterator_category = std::input_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = async_generator::value_type;
+    using reference = async_generator::reference_type;
+    using pointer = std::add_pointer_t<value_type>;
+
+    explicit iterator(handle_type coroutine) noexcept
+        : _coro{coroutine}
+    {}
+
+    explicit operator bool() const noexcept {
+        return _coro && !_coro.done();
+    }
+
+    [[nodiscard]] auto operator++() noexcept {
+        using base_awaiter = internal::next_awaiter<yielded_type>;
+        class increment_awaiter final : public base_awaiter {
```

Comment:
Aha, so this is how we can be sure the consumer is a coroutine.

I think we should relax this, though my brain will probably explode.

Reply:
to accommodate both asynchronous and synchronous producers, we could specialize the generator based on the 
`Promise` type passed to `std::coroutine_handle`. However, this approach would introduce additional complexity to the implementation. i recommend utilizing the standard library's generator implementation,
which offers a more comprehensive set of features.

Reply:
You mean synchronous and asynchronous consumers, no? The producer is always a coroutine.

Reply:
> You mean synchronous and asynchronous consumers, no?

ahh, sorry. i was thinking about the async generator versus sync generator. but `seastar::coroutine::experimental::generator` implementation always assumes the promise is a seastar task. To obtain elements from an instance of this generator, one must use `co_await`. consequently, it's challenging to envision a consumer of this generator that isn't itself a coroutine.

> The producer is always a coroutine.

indeed. i just realized that it's more about async versus sync not about regular function versus coroutine.